characteristic_read/web_app.js

```diff
@@ -0,0 +1,86 @@
+/**
+ *  Copyright 2020 Google LLC
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+const testServices = ['0b30acec-193e-11eb-adc1-0242ac120002'];
+
+const testCharacteristic = '0b30afd0-193e-11eb-adc1-0242ac120002';
+
+/**
+ * Load the device code to the Espruino IDE.
+ */
+function loadCode() {
+  fetch('device_code.js').then(response => response.text()).then(data => {
+    let url = 'http://www.espruino.com/webide?code=' + encodeURIComponent(data);
+    window.open(url, '_window');
+  });
+}
+
+function onGattDisconnected(evt) {
+  const device = evt.target;
+  logInfo(`Disconnected from GATT on device ${device.name}.`);
+}
+
+async function startTest() {
+  clearStatus();
+  logInfo('Starting test');
+
+  let options = { acceptAllDevices: true };
```

Comment:
This should filter on `testServices[0]` to help the person running the test find the correct device in noisy RF environments.

Reply:
I added the filter and removed the other two options. At present nothing passes the filter. I'm still looking into it, but wanted to post what I have so far.

Reply:
This is probably because testServices[0] isn't being advertised. You can modify the call to `NRF.setServices()` or filter on the Nordic UART service which the device might be advertising: "6e400001-b5a3-f393-e0a9-e50e24dcca9e"

Reply:
So what I got working was to filter on the Nordic UART service. I expected this to work, but it did not:

device:
```c++
NRF.setServices({
    '0b30acec-193e-11eb-adc1-0242ac120002': {
      '0b30afd0-193e-11eb-adc1-0242ac120002': {
        value: [17],
        broadcast: false,
        readable: true,
        writable: false,
        notify: false,
        description: 'Single read-only characteristic',
      }
    }
  }, { advertise: ['0b30acec-193e-11eb-adc1-0242ac120002'] });
```

webapp:
```js
const options = {
  filters: [
    { services: [testService] }
  ]
};
const device = await navigator.bluetooth.requestDevice(options);
```

So, I like that this test filters on advertised services, but it does make it harder to help users troubleshoot problems where the need to connect to a device with a non-advertised service.